### PR TITLE
issue#53 duplicate HASH key on a new SET_QUERY building

### PR DESCRIPTION
**Pull request by lrivera-oicr**

### CODE REVIEW

File: src/main/java/org/cancogenvirusseq/singularity/config/archive/ArchiveProperties.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.Data;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@Configuration
+@ConfigurationProperties(prefix = "archive")
+
+public class ArchiveProperties {
```
Comment (lrivera-oicr) [inline src/main/java/org/cancogenvirusseq/singularity/config/archive/ArchiveProperties.java:11]: created this new Class to group some archive properties, so can be used by any other class.

File: src/main/resources/application.yaml [hunk 2]
```diff
@@ -60,3 +60,4 @@ utils:
 
 archive:
   cancelPeriodSeconds: 14400
+  maxBuildingSeconds: 14400
```
Comment (lrivera-oicr) [inline src/main/resources/application.yaml:63]: new property to define the maximum amount of time for an archive to be build. 
Default 14400 seconds (4 hours)

File: src/main/java/org/cancogenvirusseq/singularity/components/pipelines/SetQueryArchiveRequest.java [hunk 3]
```diff
@@ -99,20 +106,39 @@ private Function<Archive, Mono<Archive>> saveAndTriggerBuildOrGetArchiveFunction
...
-                                archivesRepo.findArchiveByHashInfoEquals(archive.getHashInfo()))
-                        .orElseThrow(() -> dataViolation));
+                dataViolation -> getSqlStateOptionalFromException(dataViolation)
+                    .filter(isUniqueViolationError)
+                    .map(
+                      uniqueConstraint -> archivesRepo
+                        .findArchiveByHashInfoEquals(archive.getHashInfo())
+                        .flatMap(a -> {
+                          if(ArchiveStatus.CANCELLED.equals(a.getStatus()) ||
+                            (ArchiveStatus.BUILDING.equals(a.getStatus()) && a.getCreatedAt() < Instant.now().minusSeconds(archiveProperties.getMaxBuildingSeconds()).getEpochSecond())){
```
Comment (caravinci) [inline src/main/java/org/cancogenvirusseq/singularity/components/pipelines/SetQueryArchiveRequest.java:120]: Think I would visually group the different parts of that `|| &&` combo, so it's clearer which conditions go with which. Also `a`? that variable name could be more declarative (e.g.`archive` or `newArchive`)

Comment (lrivera-oicr) [reply src/main/java/org/cancogenvirusseq/singularity/components/pipelines/SetQueryArchiveRequest.java]: moved this condition logic in to a new method.

**Review by caravinci — APPROVED** [review]
Looks great!

**Pull request merged**